Add a warning log when there is high skew of uneven inputs in DDP training

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45238 Add a warning log when there is high skew of uneven inputs in DDP training**

This request came up in feature review for DDP uneven inputs, so this PR adds a warning when there is much higher than expected amount of
discrepancy of inputs across different processes when running with uneven
inputs. This is because a skew in the thousands can reduce performance a
nontrivial amount as shown in benchmarks in https://github.com/pytorch/pytorch/pull/42577, and it was proposed to add this
warning as a result. Tested by running the tests so the threshold is hit and
observing the output.

Differential Revision: [D23719270](https://our.internmc.facebook.com/intern/diff/D23719270/)

rohan-varma committed:
Add a warning log when there is high skew of uneven inputs in DDP training

Adds a warning when there is much higher than expected amount of
discrepancy of inputs across different processes when running with uneven
inputs. This is because a skew in the thousands can reduce performance a
nontrivial amount as shown in benchmarks, and it was proposed to add this
warning as a result. Tested by running the tests so the threshold is hit and
observing the output.

Differential Revision: [D23719270](https://our.internmc.facebook.com/intern/diff/D23719270/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add a warning log when there is high skew of uneven inputs in DDP training"


This request came up in feature review for DDP uneven inputs, so this PR adds a warning when there is much higher than expected amount of
discrepancy of inputs across different processes when running with uneven
inputs. This is because a skew in the thousands can reduce performance a
nontrivial amount as shown in benchmarks, and it was proposed to add this
warning as a result. Tested by running the tests so the threshold is hit and
observing the output.

Differential Revision: [D23719270](https://our.internmc.facebook.com/intern/diff/D23719270/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add a warning log when there is high skew of uneven inputs in DDP training"


This request came up in feature review for DDP uneven inputs, so this PR adds a warning when there is much higher than expected amount of
discrepancy of inputs across different processes when running with uneven
inputs. This is because a skew in the thousands can reduce performance a
nontrivial amount as shown in benchmarks, and it was proposed to add this
warning as a result. Tested by running the tests so the threshold is hit and
observing the output.

Differential Revision: [D23719270](https://our.internmc.facebook.com/intern/diff/D23719270/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Add a warning log when there is high skew of uneven inputs in DDP training"


This request came up in feature review for DDP uneven inputs, so this PR adds a warning when there is much higher than expected amount of
discrepancy of inputs across different processes when running with uneven
inputs. This is because a skew in the thousands can reduce performance a
nontrivial amount as shown in benchmarks in https://github.com/pytorch/pytorch/pull/42577, and it was proposed to add this
warning as a result. Tested by running the tests so the threshold is hit and
observing the output.

Differential Revision: [D23719270](https://our.internmc.facebook.com/intern/diff/D23719270/)

[ghstack-poisoned]

mrshenli commented on torch/nn/parallel/distributed.py:
```diff
@@ -815,8 +815,21 @@ def join(self, divide_by_initial_world_size=True, enable=True):
+                            "other currently active ranks. This level of skew could "
+                            "lead to performance degradation during training."
+                        )
+                        warned = True
```
Please feel free to ignore. This can also be done using `warnings.simplefilter("once")` IIUC.

rohan-varma replied:
Thanks! This is better than toggling the boolean.

mrshenli reviewed (approved):
LGTM!

rohan-varma committed:
Update on "Add a warning log when there is high skew of uneven inputs in DDP training"


This request came up in feature review for DDP uneven inputs, so this PR adds a warning when there is much higher than expected amount of
discrepancy of inputs across different processes when running with uneven
inputs. This is because a skew in the thousands can reduce performance a
nontrivial amount as shown in benchmarks in https://github.com/pytorch/pytorch/pull/42577, and it was proposed to add this
warning as a result. Tested by running the tests so the threshold is hit and
observing the output.

Differential Revision: [D23719270](https://our.internmc.facebook.com/intern/diff/D23719270/)

[ghstack-poisoned]
